File: peer_logger.py
import threading
from datetime import datetime


class PeerLogger:
    def __init__(self, peer_id):
        self.peer_id = peer_id
        self._lock = threading.Lock()
        self._file = open(f"log_peer_{peer_id}.log", "w")

    def _timestamp(self):
        return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    def _write(self, msg):
        with self._lock:
            self._file.write(msg + "\n")
            self._file.flush()

    def log_tcp_connection_to(self, remote_id):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} makes a connection to Peer {remote_id}.")

    def log_tcp_connection_from(self, remote_id):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} is connected from Peer {remote_id}.")

    def log_preferred_neighbors(self, neighbor_ids):
        ids = ", ".join(str(n) for n in neighbor_ids)
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} has the preferred neighbors {ids}.")

    def log_optimistically_unchoked_neighbor(self, neighbor_id):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} has the optimistically unchoked neighbor {neighbor_id}.")

    def log_unchoking(self, remote_id):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} is unchoked by {remote_id}.")

    def log_choking(self, remote_id):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} is choked by {remote_id}.")

    def log_have(self, remote_id, piece_index):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} received the 'have' message from {remote_id} for the piece {piece_index}.")

    def log_interested(self, remote_id):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} received the 'interested' message from {remote_id}.")

    def log_not_interested(self, remote_id):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} received the 'not interested' message from {remote_id}.")

    def log_downloaded_piece(self, remote_id, piece_index, total_pieces):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} has downloaded the piece {piece_index} from {remote_id}. Now the number of pieces it has is {total_pieces}.")

    def log_download_complete(self):
        self._write(f"[{self._timestamp()}]: Peer {self.peer_id} has downloaded the complete file.")

    def close(self):
        with self._lock:
            self._file.close()

File: PeerLogger.java
import java.io.*;
import java.text.SimpleDateFormat;
import java.util.*;

public class PeerLogger {
    private PrintWriter writer;
    private int peerId;

    public PeerLogger(int peerId) {
        this.peerId = peerId;
        try {
            writer = new PrintWriter(new FileWriter("log_peer_" + peerId + ".log"), true);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private String getTimestamp() {
        return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());
    }

    public synchronized void logTCPConnectionTo(int remotePeerId) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " makes a connection to Peer " + remotePeerId + ".");
    }

    public synchronized void logTCPConnectionFrom(int remotePeerId) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " is connected from Peer " + remotePeerId + ".");
    }

    public synchronized void logPreferredNeighbors(List<Integer> neighbors) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < neighbors.size(); i++) {
            if (i > 0) sb.append(", ");
            sb.append(neighbors.get(i));
        }
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " has the preferred neighbors " + sb.toString() + ".");
    }

    public synchronized void logOptimisticallyUnchokedNeighbor(int neighborId) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " has the optimistically unchoked neighbor " + neighborId + ".");
    }

    public synchronized void logUnchoking(int remotePeerId) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " is unchoked by " + remotePeerId + ".");
    }

    public synchronized void logChoking(int remotePeerId) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " is choked by " + remotePeerId + ".");
    }

    public synchronized void logHave(int remotePeerId, int pieceIndex) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " received the 'have' message from " + remotePeerId +
                " for the piece " + pieceIndex + ".");
    }

    public synchronized void logInterested(int remotePeerId) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " received the 'interested' message from " + remotePeerId + ".");
    }

    public synchronized void logNotInterested(int remotePeerId) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " received the 'not interested' message from " + remotePeerId + ".");
    }

    public synchronized void logDownloadedPiece(int remotePeerId, int pieceIndex, int totalPieces) {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " has downloaded the piece " + pieceIndex + " from " + remotePeerId +
                ". Now the number of pieces it has is " + totalPieces + ".");
    }

    public synchronized void logDownloadComplete() {
        writer.println("[" + getTimestamp() + "]: Peer " + peerId +
                " has downloaded the complete file.");
    }

    public synchronized void close() {
        if (writer != null) {
            writer.close();
        }
    }
}
